### DISCUSSION TITLE
Provide option to allow tests to be ran before a model's data is updated

### DISCUSSION THREAD

**Discussion by adamcunnington-mlg** [dbt-labs/dbt-core#5687] [Ideas]
# Describe the feature

## Context
Currently, DBT tests run after a model is built.
If a test fails, there is some characteristic of the data that is unexpected - but it's too late. That "bad" data is in the model and is impacting downstream models - the integrity of the data is broken and this could flow through to other models where crazy things might happen (row explosion uh oh) and use cases (i.e. real-time dashboards) suddenly show the wrong data.

There is no straight forward solution to this problem. On snowflake, there is a neat way of achieving blue/green (although I think it's a slight misuse of the term blue/green because we're talking about new data, not [necessarily] new software) because Snowflake supports partition swaps. Others have implemented more complex workarounds such as managing separate staging and production environments - orchestrated either via Airflow, or perhaps achieved in DBT via logical layers with different run commands. Calogica have a nice write-up about how they use DBT to achieve this [here](https://calogica.com/assets/wap_dbt_bigquery.pdf).

**My conjecture though is that this is a really core need that should be possible, natively, within DBT.**

## Acknowledgements
I am aware that there are use cases where you don't want to prevent "bad" data flowing through in the event of a test failure - but I do believe it should be an option - and should probably be the default behaviour (ignoring for one moment compromises that might be made when rolling out such a feature as to not cause a breaking change).

Also, it's worth calling out that tests actually serve two purposes:
1) Validating a change in model configuration / SQL has not caused regression
2) Validating that new data meets the expected profile

These are arguably quite different things. The former is more like a conventional software test and we'd expect to run that before deploying our updated artifacts (in this case, updated model code) - which we do via a typical DBT dev workflow. The second is a runtime consideration relating to the assertion of our data's profile. This is a very different thing and why it's so important to be able to succeed the tests before updating the model's data (or at least have the option to do this). The first is a solved problem. The second is not.

## Discussion
DBT makes decisions about what should be abstracted behind a common layer and what should be adapter-specific. It feels to me like the ability, and user-facing interface, for having tests run (logically) before a model is updated, should be adapter-agnostic - but I appreciate this is a totally non-trivial problem to solve. Perhaps adapter-specific could be a starting point and this could be homogenised later.

I'm also aware, at the other end of the spectrum, that if you are going to consider running tests "transactionally" before a model is updated, what about a model that is dependent on another model? You may wish to update both or none at all. Sure! But one step at a time.

I imagine a world where there is a per-adapter implementation but perhaps hidden behind a common model property which controls whether tests run before or after.

Specifically, and selfishly, I want this capability in BigQuery. BigQuery is quite limited in terms of tools it gives us to help but multi-statement transactions have been round a shortwhile and are due to move from preview to GA in the not too distant future. I know transactions are not a silver-bullet solution to everything but maybe there's at least some mileage in exploring the feasibility of updating model and then running all model & column-level tests within the same transaction and rolling back if any fail (maybe the failure behaviour is a config that belongs to the `tests:` array items rather than at the model level)?

Keen to start a discussion on this!

# Describe alternatives you've considered

Managing a custom blue/green approach.

# Who will this benefit?

A lot of people. I don't have any hard numbers to back this up but I suspect that the **majority** of DBT users would not expect a model to be updated if tests fail. I imagine some of those users implement workarounds which cost development time and processing resources and opportunity cost in complex support routines. I imagine that others don't do anything in particular but just set expectation with users (directly or otherwise) that when data breaks, it'll be fixed quickly - but it will be broken for a period of time. Most data analytics still isn't mission critical after all (despite what we all say).

# Are you interested in contributing this feature?

Perhaps! Certainly dev resources within my team.

# Anything else?

https://cloud.google.com/bigquery/docs/reference/standard-sql/transactions

**Comment by rileyschack** [comment]
+1 for this ability. Realize it may not be the easiest implementation. My org is starting to experiment with table/dataset cloning as a means to work with dbt's current functionality. Even contemplating building a clone materialization that would only run if all tests are passed (using `dbt build`).

**Comment by dbeatty10** [comment]
#1054 has an interesting discussion on this topic.

**Comment by jtcohen6** [comment]
@adamcunnington-mlg This is a tremendous start to a discussion! You've compiled some great research here. I am going to convert this to a discussion, in matter of fact.

I also agree with @dbeatty10 that #1054 offers some great historical context on the ideas we've had in the past, as well as the tripwires we've come across.

I appreciate the problem you're raising—when bad data's in the production pipeline, _it's in the production pipeline_—and I agree it's a very common one. There are a diversity of tactics folks use when solving it today. The two most common I see:
1. **Define tests on `sources`**, to validate that new data meets the expected profile. In the unified DAG of `dbt build`, those tests will run _before_ any models depending on those sources are built, and those models will skip if the tests fail.
2. **Use a "write-audit-publish" or "blue/green" pattern** to run, test, and only then deploy models to production. You can do all three steps once per model, although the pattern I see more often is to swap an entire schema/database from "staging" to "production" when it's ready. This _also_ avoids the risk of downstream queriers seeing inconsistent data, if one "final" model finishes building earlier than another "final" model.

While not perfect solutions, those two constructs have served many quite well. They care more about the holistic deployment of an entire DAG, rather than treating each model as its own microcosm.

I do think that dbt could do more to make the write-audit-publish / "blue-green" pattern more readily available. It is certainly easiest on Snowflake, due to its support of zero-copy cloning, though still requires some custom macros + operations to string together. BigQuery also supports zero-cost copying, for tables and datasets, and (I understand) is rolling out capabilities around [table cloning](https://cloud.google.com/bigquery/docs/table-clones-intro) (for discussion in https://github.com/dbt-labs/dbt-bigquery/issues/270). Once there is a critical mass of support among a few of the most popular adapters, I feel more comfortable with the idea of building out an adapter-agnostic abstraction within dbt.

---

## Each model to its own

There are three patterns I could see to support this on a model-by-model basis, all of which are much more case-by-case:
1. **Tests within transactions:** On data platforms that support transactions, run tests before the `commit;` and roll back if they fail. This is totally adapter-specific: Redshift supports transactions completely, Spark/Databricks not at all, Snowflake in very odd ways, and BigQuery only very recently. Within dbt, it would require us to rethink tests as a property of a model, which run during that model's materialization, rather than their own node types within the DAG. (How to treat tests that select from multiple models...?) I'm not dismissing this idea completely, just noting that results will vary significantly here, and I'd want to take account of the several adapter/materialization permutations before proceeding.
2. **Tests within materializations:** For certain materializations on certain adapters (namely `incremental`), which produce a temporary table before upserting/merging/overwriting into the existing result set, test the temporary results _before_ upserting/merging/overwriting. This is an issue that Niall just opened, and to which I owe a response, though his motivation is more about performance of the test queries than rolling back if the tests fail: https://github.com/dbt-labs/dbt-core/issues/5427. The appeal of this approach doesn't strictly require transactions, because the temp table can be thrown out without a formal rollback—no harm, no foul—but it does get into the implementation details of specific materializations on specific adapters, and it raises the same questions about how tests need to be defined in the manifest and run in the DAG.
3. **Database constraints:** Historically, analytical databases have done something very ugly with constraints: they allow them, they _use_ them in query optimization, but they do not _enforce_ them. So, despite adding a `unique` constraint to a column, it may well have dupes—and the database's optimizer will act as if it _doesn't_, risking incorrect results! For that reason, dbt has really, really discouraged use of database constraints in the past. Still, there is a pattern of cross-applying dbt tests as database constraints for detection by other tools (https://hub.getdbt.com/snowflake-labs/dbt_constraints/latest/). On at least one data platform (Databricks), constraints are _totally_ enforced at table creation/update time (https://github.com/databricks/dbt-databricks/pull/71). That feels like a pretty good pattern, where it's available!

---

I agree that, in the ideal case, there would be a common abstraction in dbt to identify which tests are worth rolling back for—with adapter-specific implementations behind the scenes—just as `dbt build` has allowed users to say that all error-severity tests are worth stopping the DAG for. I'm also wary of building an abstraction that risks leaking beyond utility, because of how different these data platforms really are, behind their shiny SQL-y exteriors.

Let's keep the discussion going :)

**Comment by adamcunnington-mlg** [comment]
@jtcohen6 thanks for the response - exactly the sort of discussion I was hoping to generate!

I hadn't fully appreciated (despite the fact that I am actually benefiting from the exact behaviour you described) that the models are dependent on the sources and so source tests do provide a significant safety barrier. It doesn't cause the discussion to be moot as we can still miss data issues through not catching edge cases in our source tests and then model tests may fail (and it's too late) but I think it would be good to increase the emphasis of this in docs. Somewhere during the tutorial / getting started pages, it could be really clear that tests run after models but the way dbt build works is such that models will not run at all if source test fails and thus source tests are a critical opportunity to verify upstream data integrity. This reduces the job of model tests to then be i) validating the model logic is good - which is part of typical software dev/test workflow and falls into the solved problem 1 category I outlined before, and ii) catching edge cases that are pertinent to the logic of the model - which should drive a feedback loop in identifying gaps in source tests and improving them.

In fact, the above leads me to think that it would be a valid conclusion for the original hope of this discussion to be met with "it's not feasible given the amount of effort, and that source tests run first" - this is not something I had appreciated at all at the top of this conversation.

Onto the options though.. thanks for listing them out - I guess as much as possible, for sake of code maintainability (and accommodating future adapters), you want as few strategies as possible for achieving this - some of what you have listed are broad-brush strategies and others (e.g. constraints) are specific to particular types of tests, nevermind particular adapters. I guess leaning on the former as much as possible (even with limitations) may be preferable. I.e. if any dev work to be done, would a good starting point be to implore the transaction rollback approach behind a common model property that only has an effect on adapters that support transactions, and just call this out as a limitation? It feels like it would be the smallest code change too. Logically:
1) A model property which controls whether tests run in a transaction or not.
2) If so, and if supported by adapter, wrap the existing model running + test running inside of transaction control statements, and handle the test errors.

**Reply by adamcunnington-mlg** [reply]
Further, I also hadn't appreciated that it is already the behaviour of `dbt build` to not run model B (dependent on model A) if model A's test fails. Only model A output is bad.

When I tried to verify, no one called these things out! The existing behaviour significantly reduces the priority of this discussion item in my opinion. Basically, the only thing that DBT leaves itself a bit exposed on is a single given model showing "bad" data following a test fail. Downstream models are protected. And the fact that all sources run before models means if the source test coverage is good, you're even more protected.

If I'd have known this, I probably wouldn't have even raised this item - I'd have just accepted it as a limitation. I'm glad that I misunderstood because I still think it's a worthy enhancement to run tests within a transaction - but for sure, I'd have a dedicated page in docs - or at least a big FAQ which describes in exactly what situation a model can be at risk of being "bad" and the workaround to minimise this risk. @jtcohen6 I think this is really the only short-term ask from this discussion.

Thanks!

**Reply by jtcohen6** [reply]
@adamcunnington-mlg I'm really glad you opened the discussion! We've now got a conversation, full of valuable and recent context, that will come up in search engines. It's a really legit way of contributing to the knowledge loop.

If you've got a specific idea of where in the docs that sort of section/FAQ should live—[tests intro](https://docs.getdbt.com/docs/building-a-dbt-project/tests)? [`build` command](https://docs.getdbt.com/reference/commands/build)?—the docs site is open source as well, and you can [open an issue](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose) / PR proposing to improve the content there.

**Reply by jsnb-devoted** [reply]
>Further, I also hadn't appreciated that it is already the behaviour of dbt build to not run model B (dependent on model A) if model A's test fails. Only model A output is bad.

I do feel like `build` or blue/green deployments only solve this problem if you are running your models as 1 contiguous graph DWH run (eg "I run all my models nights"). I know a lot of people do this but our deployment has thousands of models that we execute in groups at different times because areas of the business need their data refreshed at different intervals. We implemented a macro that does exactly what @rileyschack mentioned above:
>Even contemplating building a clone materialization that would only run if all tests are passed (using dbt build).

I'm sure it has been floated elsewhere so feel free to point me to the relevant thread but what would be the downside of this model execution strategy?:
1. Create a "swap" table (eg `my_model_tmp`) based on the metadata in the schema.yml file
2. Run the model's compiled SQL as an INSERT statement into `my_model_tmp`
3. Execute the dbt tests against `my_model_tmp`
4. If the tests pass swap the tables with either `DROP` + `ALTER TABLE RENAME` or `CREATE OR REPLACE my_model AS SELECT * FROM my_model_tmp` (the latter maybe being more costly for BigQuery or Snowflake users)

This would still work for folks that do the "nightly" style DWH dbt builds but it would also work for those of us that have a less "monolithic" data pipeline environment. This would require the `schema.yml` file to have all the column names/types. I feel like that might be controversial but I do think there is a lot of benefit in having that metadata documented as a requirement.

**Reply by kparaju** [reply]
@jsnb-devoted 

> 1. Create a "swap" table (eg my_model_tmp) based on the metadata in the schema.yml file
> 2. Run the model's compiled SQL as an INSERT statement into my_model_tmp
> 3. Execute the dbt tests against my_model_tmp

Forgive me if this is a stupid question (I'm not a Data Engineer, just adjacent to it) but will this work if the model is incremental? We might need to copy `my_model` to `my_model_tmp`, right?

Use-case: What if one of the tests calculates the avg across the whole entire table? In your proposal, we will only be doing tests against the "incremental" data that is located in `my_model_tmp`, not all of the data in `my_model`. This could cause a situation where the tests pass in my_model_tmp but fail when it's run on my_model.

**Reply by jsnb-devoted** [reply]
@kparaju -- not a stupid question at all. Incremental models will have to look a little different. They would have to have a separate "materialization" but that isn't different from how dbt works now. Incremental models and a "swap table" strategy could work something like this:
1. Create a copy of the existing target table `my_inc_model_tmp`
2. Create a table for the incremental data to go `my_inc_model_staging`
3. Execute the `merge` or `delete+insert` against `my_inc_model_tmp`
4. Run your tests against `my_inc_model_tmp` and if the tests pass...
5. Drop `my_inc_model`
6. Rename `my_inc_model_tmp` to `my_inc_model`

**Reply by kparaju** [reply]
@jsnb-devoted That makes sense, thank you.

> the latter maybe being more costly for BigQuery or Snowflake users

+1; the cost will scale with the data in the target table for BQ/SF and this will cause significant compute-debt long term. Improvements like https://github.com/dbt-labs/dbt-bigquery/issues/270 will be essential to enabling this feature so as to not skyrocket the compute cost for existing users of `dbt test`. 

And echoing:
> I still think it's a worthy enhancement to run tests within a transaction 

Maybe this can be generalized at the DB level. Something like a multi-session transaction. We would:
1. Start a session (dbt run), create a transaction, get an ID
2. Make changes to the target table in that session
3. Close the session (dbt run finishes)
4. Start another session for (dbt test) using the transaction ID from (1)
5. Test
6. COMMIT or ROLLBACK based on the test result

**Reply by tema-popov** [reply]
I would like to have the ability to check assertions before running the model, in the same way, DBT test work for post-model run checks. 

I have two cases and they all affect specifically large incremental models:
1. I need to check the state of the incremental model before updating it to check the consistency of updated data.
2. I need to check if the incremental model has all expected increments already built to have all the data for another model that depends on it. For example, I have a "stats for last 7 days" model that expects the incremental "1-day stats model" to have all increments for the last 7 days built.

**Reply by jsnb-devoted** [reply]
> +1; the cost will scale with the data in the target table for BQ/SF and this will cause significant compute-debt long term.

Ya I think the CREATE TABLE AS SELECT... isn't really an efficient solution. Not really sure why I included it in retrospect. It really ought to be DROP TABLE and RENAME or even better use CLONE if your db supports it.

> Maybe this can be generalized at the DB level. Something like a multi-session transaction.

Ultimately I think this is probably the right solution and more generalizable across different databases. We are using Snowflake and it hasn't always treated transactions the way I would expect. If we were doing something like this in a transactional db like postgres it would be a no brainer. Might still be the way to go with these MPP dbs.

**Comment by jason-albrecht** [comment]
Not sure if this is exactly related, but:

Being able to run newly developed tests and macros from a user's local machine against an environment where they don't have the rights to build models (create tables), would be extremely helpful.

Right now, if a user doesn't have CREATE TABLE rights, when they `dbt test` a model that does not have a new column yet (because it's being developed locally), the test fails.

Running the local dbt project in a temp environment against real data would be amazingly useful.

**Reply by adamcunnington-mlg** [reply]
Hi @jason-albrecht, it's not related.
You are talking about a very common need and it's covered by https://docs.getdbt.com/reference/node-selection/defer

**Comment by SoumayaMauthoorMOJ** [comment]
@adamcunnington-mlg very useful discussion! With the "run, test, and only then deploy models to production"  approach, models downstream of a failed test will run needlessly. However with `dbt build`, models downstream of a failed test would not run. Another issue is that **all** tests need to pass before deploying to production. If any test fails, no tables will be published to production. With `dbt build`, **only** models downstream of a failed test would not run.

Since `dbt build` knows that it needs to run models and tests in the right order, how about modifying `dbt build` so that you can optionally run something like `on-test-end`,  similar to 'on-run-end', but which runs after tests have completed for a particular model. You could pass different macro depending on whether test passed or failed, for example to swap the database / partition, or roll-back to a previous [branch](https://iceberg.apache.org/docs/latest/branching/) whatever fits your use case and/or adapter.

**Reply by SoumayaMauthoorMOJ** [reply]
I’ve created a [dbt](https://github.com/moj-analytical-services/jaffle_shop_duckdb_wap) project to demonstrate how to implement WAP on table [materializations](https://docs.getdbt.com/docs/build/materializations) using “dummy” post-test-hooks:

1. Create a ‘wap’ version of your model:
```
--my_table_wap.sql
{{ config(
    table_type='iceberg',
    s3_data_naming='unique',
    s3_data_dir=transform_table.generate_s3_location(),
) }}

SELECT 1 AS id
```

2. Create a view model that references the wap model, and then deletes the view and renames the wap table as post-hook:
```
--my_table.sql
{{ config(
    materialized='view',
    post_hook=[ 
        "DROP VIEW {{ this }}",
        "{{ rename_relation(ref('my_table_wap'),this) }}",
    ],
) }}

SELECT * FROM {{ref('my_table_wap')}}
```

It applies the approach to the example [jaffle_shop_db](https://github.com/dbt-labs/jaffle_shop_duckdb) dbt project which uses the [dbt-duckdb](https://github.com/duckdb/dbt-duckdb) adapter. 

What do you think of this approach? Can you think of any cons? Do you know how to enable this pattern in dbt-core without the need for the “dummy” post-test-hooks?

**Reply by SoumayaMauthoorMOJ** [reply]
One con with this approach is that I can't run something if my test fails e.g. to rollback a change

**Reply by SoumayaMauthoorMOJ** [reply]
It's also incompatible with incremental models

**Reply by SoumayaMauthoorMOJ** [reply]
I've come up with a WAP approach that works for incremental models, but only for table that have time travel querying capabilities e.g. Iceberg:

1. Create a ‘wap’ incremental version of your model:
```
--my_table_wap.sql
{{
    config(
        materialized='incremental',
        table_type='iceberg',
        incremental_strategy='append',
    )
}}

SELECT 1 as id
  
```

2. Create a view model that references the wap model but as of a set [TIMESTAMP](https://docs.aws.amazon.com/athena/latest/ug/querying-iceberg-table-data.html)
```
--my_table.sql
{{
    config(
        materialized='view'
    )
}}

{% set current_timestamp = modules.datetime.datetime.now() %}

SELECT * 
FROM {{ ref('my_table_wap') }}
FOR TIMESTAMP AS OF TIMESTAMP '{{ current_timestamp }} UTC'
```

The view will only get refreshed to use the latest Iceberg table version if the tests passes

**Reply by y0j0** [reply]
awesome! This idea helped me a lot. But I found one case that it can't solve - when the tests fail during the `--full-refresh`, the snapshot used in the VIEW does not exist anymore because during the --full-refresh all old snapshots are deleted. Any possible solution that I found is to do the --full-refresh each time on the new schema. Or do you have any other ideas please?

**Reply by SoumayaMauthoorMOJ** [reply]
Create a macro to delete all the data from an iceberg table which only runs in incremental mode and if overwrite = True

```
{% macro delete_from() %}

    {% if is_incremental() and config.get('overwrite', false) %}

            DELETE FROM {{ this }}
            {%- set query -%}
                deleted from {{ this.render_pure() }}
            {%- endset -%}
            {% do log(query, info=True) %}

    {%- endif -%}

{% endmacro %}
```
 
You can then call this macro in incremental models a pre-hook to delete all the data instead of doing a full refresh:

```
{{
    config(
        materialized ='incremental',
        incremental_strategy ='append',
        table_type ='iceberg',
        pre_hook ="{{ transform_table.delete_from() }}",
        table_properties = {'vacuum_max_snapshot_age_seconds':1},
        overwrite = var('OVERWRITE', false)
    )
}}

{% if not is_incremental() or config.get('overwrite', false) %}

    SELECT 2 AS id

{% else %}

    SELECT 1 AS id

{%- endif -%}
```

To run this model in overwrite mode you'll need to use vars: `dbt run --vars 'OVERWRITE: true'`

**Comment by Spince** [comment]
wondering if dbt clone changes things here.. thinking (similar to others here):
- clone production data
- run and test against clone
- if passed: swap updated clone with production
- drop clone 

my current team doesn't have a pager and failing jobs can sometimes go days without being noticed. seems like this is a totally doable implementation especially given the benefit of only having tested data in production. I see this as a logical next step in line with the shift from dbt run to dbt build.

**Reply by mrcool4** [reply]
I believe this only solves half of the problem, what about the 2nd problem?
"Validating that new data meets the expected profile"

**Comment by larruda** [comment]
I was hoping a behavior more like https://docs.databricks.com/aws/en/dlt/expectations (operates at row level)
I'd be able to specify whether rows should be dropped or not depending on the data test given (e.g not null).

**Reply by mroy-seedbox** [reply]
- Fail + Warn: Use a data test which produces a warning instead of an error.
- Fail + Drop: Exclude those records in your model (e.g. `WHERE column IS NOT NULL`).
- Fail + Fail Flow: Use a data test which produces an error.
  - The data test can be configured on the upstream model/source in order to prevent the downstream steps from running.

**Comment by mroy-seedbox** [comment]
Apologies for the bluntness, but to me this entire feature request makes no sense at all. Data should be tested upstream, period. No need to add an entirely new feature. There are far too many ways to write data into a table for DBT to know how to test that data beforehand (especially all the various incremental strategies and custom materializations).

Data tests are a standalone utility which run on the data in a table, _and are completely decoupled from how the data in that table got there in the first place, which is a good thing_.

If you want to validate that new data meets an expected profile before writing it into a table, then add the necessary upstream models and configure tests on those models, which will prevent bad data from flowing downstream.

Those upstream models could even just be views or temporary tables. And to avoid duplicating logic, your downstream model can just select from those new upstream models (i.e. `SELECT *`) instead of repeating the same query. Or just swap the tables. There are many ways to do this depending on the needs. It is much too variable to offer a universal solution.

Basically, instead of:
- `A --> B`, and run some tests on `B`, _but before writing the data_.

You can just do:
- `A --> X --> B`, and run your tests on `X` (which will run before writing to `B`).
- `X` could even just be a clone of `B` (which could be created/updated in a [pre-hook](https://docs.getdbt.com/reference/resource-configs/pre-hook-post-hook) and then optionally dropped in a post-hook or [after the run](https://docs.getdbt.com/reference/project-configs/on-run-start-on-run-end)).
- And the model for B could just be `ALTER TABLE B SWAP WITH X;`, _which would only happen if the tests on X succeeded_.

**Reply by jsnb-devoted** [reply]
Maybe you do agree? 

You effectively are describing the essence of the feature request here:
> You can just do:
>
> - A --> X --> B, and run your tests on X (which will run before writing to B).
> - X could even just be a clone of B (which could be created/updated in a [pre-hook](https://docs.getdbt.com/reference/resource-configs/pre-hook-post-hook) and then optionally dropped in a post-hook or [after the run](https://docs.getdbt.com/reference/project-configs/on-run-start-on-run-end)).
> - And the model for B could just be ALTER TABLE B SWAP WITH X;, which would only happen if the tests on X succeeded.

The request is that this be a part of the framework because without it -- a failed test indicates that your bad data is already being propagated downstream. 

https://www.getdbt.com/blog/testing-is-not-enough-transforming-data-quality-with-write-audit-publish

**Comment by larruda** [comment]
@mroy-seedbox, thanks for the bluntness and the detailed walkthrough of the `A -> X -> B` pattern.

While testing upstream is foundational for preventing bad **input** data, the core premise that it satisfies *all* data validation needs is incorrect and fails to account for real-world business requirements at scale.

### 1. The Critical Gap: Downstream Logic Validation

The problem we are solving is not "bad input data," but **faulty transformation logic** in the final model itself. This is a crucial distinction:

* **Upstream Test:** Assures the data **entering** the model is sound.
* **WAP Pattern (Requested Feature):** Assures the data **exiting** the transformation (the final logic) is sound.

The claim that **dbt Unit Tests** should cover logic validation is also insufficient when the test must be against **live data profiles** and **dynamic thresholds**.

For example, our final model calculates a **7-day rolling average sales volume**. A critical business validation is that this value must **never be 20% lower** than the previous 30-day average.

A Unit Test can only confirm the rolling average *function* works on static sample data. Only a **dbt Data Test** run against the final, complete model output can flag a dynamic breach of this moving threshold.

Since this error is created **by the model's own logic and its current data state**, running the test *after* the build means the bad data is already in production.

### 2. The Case for a Native Framework Feature

You correctly describe the solution:

> You can just do: `A --> X --> B`, and run your tests on **X** (which will run before writing to **B**).
> And the model for B could just be `ALTER TABLE B SWAP WITH X;`...

The fact that the solution requires users to manually write complex, adapter-specific boilerplate macros (like the `CLONE` or `SWAP` logic discussed in this thread) **proves** the need for a common abstraction.

For large organizations, this is the difference between:
a) Maintaining one complex, adapter-specific macro per team across dozens of repositories.
b) Using a single, standardized `materialized='write_audit_publish'` property supported by dbt Core.

### 3. Final Note on Rigidity

With respect, affirming that "Data should be tested upstream, **period**" suggests an unfamiliarity with the complexity and variety of business requirements found in large enterprises. For example, working at one of the largest global food and beverage manufacturers, I can assure you that **not every logic-based requirement can be pushed back to the source**.

The WAP pattern is simply a necessity for safely validating complex, downstream business logic before publishing data to mission-critical consumers.

**Reply by mrcool4** [reply]
That's very well explained, @larruda . We are exactly trying to solve the same problem.

**Reply by mroy-seedbox** [reply]
I'm not saying to only do data validation on the source. That's just ignoring the problem entirely.

What I'm saying to push the logic (transformation _and_ tests) one layer upstream and do the validation on that model before sending the data further downstream. Or adding an another intermediary layer downstream if you prefer to see it that way, it's the same end result either way (`A --> X --> B` or `A --> B --> X`, either way the final downstream layer will not accept bad data from upstream, and basically `B` & `X` are effectively copies of each other once the process is complete). It does not require swapping or anything else at all. It's like a cupbearer tasting the wine before giving the same wine to the king. Or a "guard/shield" model, if you will. We use that pattern especially for accounting data and other sensitive things which cannot be modified after the fact (some of them even have a human in a loop and/or a time delay). Swapping is just an alternative way to do the same thing without having to manage the minimal overhead of an additional model/layer (but adding a different kind of overhead to handle the swapping).

-----

There is an impassable conflict between `materialized='write_audit_publish'` and all of the other behaviors and materializations that DBT supports. You cannot have both. Supporting `materialized='write_audit_publish'` would mean giving up on the rest.

It would have to be a different configuration in order to give more flexibility to the implementation and more ability for people to choose or even create their own strategy (something along the lines of custom macros that hook in the process: `pre_transform` (which is basically [`pre-hook`](https://docs.getdbt.com/reference/resource-configs/pre-hook-post-hook)), `post_transform()` (which is basically [`post-hook`](https://docs.getdbt.com/reference/resource-configs/pre-hook-post-hook)), `post_test()`, `generate_alternate_alias_name()`, etc.).

Something like:
```yml
materialized: table | view | incremental | ephemeral | materialized view | ? | custom (ex: merge | scd | vault | etc.)
write_audit_publish:
  strategy: clone_swap | blue_green | write_twice | ? | custom
  alternate_alias: my_table_tmp # Optional
  ?: ?
```

And each `write_audit_publish` strategy would basically be a bundle of the hooks described above.

At that point, all that DBT would have to do for `clone_swap` is:
- Create a temporary clone of the table using `alternate_alias` (for systems that support cloning).
- Once the data is loaded into the clone and the tests are passing _(on the clone),_ swap the tables.
- But this might cause issues for some customers given [the amount of considerations to consider when cloning](https://docs.snowflake.com/en/user-guide/object-clone).

Another alternative for systems that do not support cloning, and which is more like a traditional blue-green deployment:
- DBT creates a view instead of a table .
- Behind the view, there are two secret tables: one `_blue` and one `_green`. The view points to the `_blue` table.
- The write happens in the `_green` table.
- If the tests pass, the tables are swapped.
- Optional: `_green` is updated again (after the swap). But it could also just be updated in the next run (depending on the use case).

The simplest implementation by far though would be to do the same write in two tables (but in the published one only after the tests pass on the audit one), so no swapping involved at all. But this only works for pipelines where no live data could be inserted upstream by a different process in between the two writes (like Snowpipe), although this can be guarded against in various different ways (including time travel). And this point we're getting pretty close to the `A --> X --> B` solution described above (but in this case it's more like `A --> X` followed by `A --> B`, _if the tests on `X` pass_, which could be implemented in an ugly way by duplicating the logic in `X` & `B`, and adding a dummy `ref('X')` in `B` -- although using a macro could avoid duplicating the logic, but that's an extra thing to manage).

So yeah, lots of different ways to do it. And I don't think that DBT should be dictating the approach. Flexibility wins.

Maybe all that's really needed is the addition of a `post_tests()` hook in order to make these operations easier/feasible.

**Reply by mroy-seedbox** [reply]
Here's a possible way to implement clone swapping without any changes to DBT (which does require the addition of another model, like `A --> X --> B` above):
- `X` has a pre-hook to clone itself from `B`.
- `X` contains the actual business/transformation logic and the tests.
- `B` swaps itself with `X` (or clones itself from `X`).
  - The `ref('X')` would ensure that `B` only runs once the tests on `X` pass.
  - Maybe it could even drop `X` in order to avoid any confusion from users seeing both tables.

Simple enough. 👌 

Now if we could just put the final swap in a `post_tests()` hook, we'd be in business! Although it would still require two models in terms of references (references being on `B` rather than `X` in order to ensure that only clean/audited/published data is served).

I think this is a pretty fair compromise, rather than implementing a whole new feature in DBT?

_**Edit:**_ I just realized that this is basically the exact same thing that I suggested in my [original post](https://github.com/dbt-labs/dbt-core/discussions/5687#discussioncomment-15032144). 🤦‍♂️ 

Sometimes it's important to say no, and if it were up to me, with such a simple workaround, I would say no to this feature request.

**Reply by mroy-seedbox** [reply]
Actually, it's even simpler: `A --> X --> B`
- `X` contains the actual business/transformation logic and the tests.
- `B` clones itself from `X` _(if the tests passed)_.

That's it.

So `B` is just a safe/tested version of `X`.

All that DBT would need to offer is a super basic `clone()` cross-platform macro (and/or model config), and nothing custom would be required. 🤷‍♂️ 

In that case, the model for `B` could just be:

```
{{ clone(this, ref('X')) }}
```

Or with a model config:

```
materialization: clone
clone: ref('X')
```

Or:

```
pre-hook: {{ clone(this, ref('X')) }}
materialization: none
```

Easy peasy. 👍 

And that `clone()` macro and `none` materialization could be really useful for many other use cases, so that would be a win for everyone.

**Reply by mrcool4** [reply]
If X is an incremental model, the above solution might not work because there is a need to rebaseline 'X' from 'B' before rebuilding 'X'.

**Reply by mroy-seedbox** [reply]
Right, every case is different, but the solution remains simple. And hopefully that rebasing is not something that needs to be done very often, so it could even just be done manually whenever needed. Or with the same simple `pre-hook` in `X`:

```
pre-hook: {{ clone(this, ref('B')) }}
```

Otherwise, how do you handle incremental models that do not use WAP? You can't rebaseline those from another model, and yet data issues still need to be handled somehow (typically by fixing the source or whatever is broken upstream and then re-running the pipeline, and making sure to tell whoever it is to stop dumping toxic waste in the river).

Either way, no bad data will ever get into `B`. The tests in `X` will make sure of it.

<img width="415" height="421" alt="image" src="https://github.com/user-attachments/assets/3b830fef-6b0c-4091-90cb-89b46a71aac1" />
